crates/oxide/src/extractor/mod.rs

```diff
@@ -0,0 +1,902 @@
+use crate::cursor;
+use crate::extractor::machine::Span;
+use candidate_machine::CandidateMachine;
+use css_variable_machine::CssVariableMachine;
+use machine::{Machine, MachineState};
+use std::fmt;
+
+pub mod arbitrary_property_machine;
+pub mod arbitrary_value_machine;
+pub mod arbitrary_variable_machine;
+pub mod bracket_stack;
+pub mod candidate_machine;
+pub mod css_variable_machine;
+pub mod machine;
+pub mod modifier_machine;
+pub mod named_utility_machine;
+pub mod named_variant_machine;
+pub mod string_machine;
+pub mod utility_machine;
+pub mod variant_machine;
+
+#[derive(Debug)]
+pub enum Extracted<'a> {
+    /// Extracted a valid looking candidate
+    ///
+    /// E.g.: `flex`
+    ///
+    Candidate(&'a [u8]),
+
+    /// Extracted a valid looking CSS variable
+    ///
+    /// E.g.: `--my-variable`
+    ///
+    CssVariable(&'a [u8]),
+}
+
+impl fmt::Display for Extracted<'_> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            Extracted::Candidate(candidate) => {
+                write!(f, "Candidate({})", std::str::from_utf8(candidate).unwrap())
+            }
+            Extracted::CssVariable(candidate) => {
+                write!(
+                    f,
+                    "CssVariable({})",
+                    std::str::from_utf8(candidate).unwrap()
+                )
+            }
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct Extractor<'a> {
+    cursor: cursor::Cursor<'a>,
+
+    css_variable_machine: CssVariableMachine,
+    candidate_machine: CandidateMachine,
+}
+
+impl<'a> Extractor<'a> {
+    pub fn new(input: &'a [u8]) -> Self {
+        Self {
+            cursor: cursor::Cursor::new(input),
+
+            css_variable_machine: Default::default(),
+            candidate_machine: Default::default(),
+        }
+    }
+
+    pub fn extract(&mut self) -> Vec<Extracted<'a>> {
+        // Candidates found by inner candidate machines. If the outer machine finds a solution, we
+        // can discard the inner machines. Otherwise, we can keep the candidates from the inner
+        // machines.
+        let mut in_flight_spans: Vec<Span> = Vec::with_capacity(15);
+
+        // All the extracted values
+        let mut extracted = Vec::with_capacity(100);
+
+        let len = self.cursor.input.len();
+
+        // CSS Variable extractor
+        {
+            let cursor = &mut self.cursor.clone();
+            while cursor.pos < len {
+                if cursor.curr.is_ascii_whitespace() {
+                    cursor.advance();
+                    continue;
+                }
```

Comment:
Maybe I could use the `fast_skip` here. Didn't even think of it.

Reply:
Yeah, possibly! I say we do this in a followup PR tho